File: Visual VB.NET/Console Applications/Sms/Gsm (with Delivery Reports)/GsmProgram.vb
'-----------------------------------------------------------------------
' <copyright file="GsmProgram.vb" company="Auron">
'     Copyright (c) Auron Software - www.auronsoftware.com
' </copyright>
'-----------------------------------------------------------------------

Imports AxSms
Imports System.IO
Imports System.Threading

Module GsmProgram

    Sub Main()
        Dim objGsm As AxSms.Gsm = New AxSms.Gsm()
        Dim objSmsMessage As AxSms.Message = New AxSms.Message()
        Dim objSmsConstants As AxSms.Constants = New AxSms.Constants()
        Dim objStatusReport As AxSms.GsmDeliveryReport = Nothing
        Dim strDevice As String, strReference As String

        ' A license key is required to unlock this component after the trial period has expired.
        ' Assign the LicenseKey property every time a new instance of this component is created (see code below). 
        ' Alternatively, the LicenseKey property can be set automatically. This requires the license key to be stored in the registry.
        ' For details, see manual, chapter "Product Activation".
        ' objGsm.LicenseKey = "XXXXX-XXXXX-XXXXX"

        ' Component info
        Console.WriteLine("Auron SMS Component {0}{1}Build: {2}{3}Module: {4}{5}License Status: {6}{7}License Key: {8}{9}", _
                          objGsm.Version, Environment.NewLine, objGsm.Build, Environment.NewLine, objGsm.Module, Environment.NewLine, _
                          objGsm.LicenseStatus, Environment.NewLine, objGsm.LicenseKey, Environment.NewLine)

        ' Set Logfile (optional, for debugging purposes)
        objGsm.LogFile = Path.GetTempPath() + "Gsm.log"
        Console.WriteLine("Log file used: {0}{1}", objGsm.LogFile, Environment.NewLine)

        strDevice = ReadDevice(objGsm)
        If (strDevice = String.Empty) Then
            Console.WriteLine("Press <ENTER> to continue.")
            Console.ReadLine()
            Exit Sub
        End If

        objGsm.Open(strDevice, String.Empty, 0)
        Console.WriteLine("Open, result: " & objGsm.LastError.ToString() & " (" & objGsm.GetErrorDescription(objGsm.LastError) & ")")
        If (objGsm.LastError = 36101) Then ' 36101 means: modem requires pin code, see also: www.auronsoftware.com/kb/api-error-codes/
            objGsm.Open(strDevice, ReadInput("Enter PIN code", False), 0)
            Console.WriteLine("Open, result: " & objGsm.LastError.ToString() & " (" & objGsm.GetErrorDescription(objGsm.LastError) & ")")
        End If

        If (objGsm.LastError <> 0) Then
            Console.WriteLine("Press <ENTER> to continue.")
            Console.ReadLine()
            Exit Sub
        End If

        ' Message settings
        objSmsMessage.ToAddress = ReadInput("Enter recipient: ", False)
        objSmsMessage.Body = ReadInput("Enter message body: ", False)
        objSmsMessage.RequestDeliveryReport = True

        ' Send the message; This sample does not work with multipart messages because the message reference would be a comma seperated
        ' string of multiple message references, one for each generated part.
        objGsm.SendSms(objSmsMessage, objSmsConstants.MULTIPART_REJECT, 0)
        Console.WriteLine("SendSms, result: " & objGsm.LastError.ToString() & " (" & objGsm.GetErrorDescription(objGsm.LastError) & ")")
        If (objGsm.LastError <> 0) Then
            objGsm.Close()
            Console.WriteLine("Press <ENTER> to continue.")
            Console.ReadLine()
            Exit Sub
        End If

        ' The message reference was set by the sendsms function on success
        strReference = objSmsMessage.Reference
        Console.WriteLine("Message Reference: " & strReference)

        ' Status Report
        Dim bSearchStatusReport As Boolean
        bSearchStatusReport = True
        Do While (bSearchStatusReport)
            objGsm.Receive(objSmsConstants.GSM_MESSAGESTATE_ALL, False, objSmsConstants.GSM_STORAGETYPE_ALL, 25000)
            Console.WriteLine("Receive, result: " & objGsm.LastError.ToString() & " (" & objGsm.GetErrorDescription(objGsm.LastError) & ")")

            If (objGsm.LastError <> 0) Then
                Exit Do
            End If

            objStatusReport = objGsm.GetFirstReport()

            Do While (objGsm.LastError = 0)
                If (objStatusReport.Reference = strReference) Then
                    bSearchStatusReport = False
                    Console.WriteLine("Status Report {0} received!", objStatusReport.Reference)
                    objGsm.DeleteReport(objStatusReport)
                    Console.WriteLine("DeleteReport, result: " & objGsm.LastError.ToString() & " (" & objGsm.GetErrorDescription(objGsm.LastError) & ")")
                    Exit Do
                Else
                    objStatusReport = objGsm.GetNextReport()
                End If
            Loop

            If (bSearchStatusReport) Then
                Console.WriteLine("Status report not found yet.")
                Thread.Sleep(3000)
            End If
        Loop

        objGsm.Close() ' Make sure the port gets closed

        Console.WriteLine("Ready.")
        Console.WriteLine("Press <ENTER> to continue.")
        Console.ReadLine()

    End Sub

    Function ReadInput(ByVal strTitle As String, ByVal bAllowEmpty As Boolean)

        Dim strInput As String
        Dim strReturn As String = String.Empty

        Console.WriteLine(strTitle)
        Do
            Console.Write("  > ")
            strInput = Console.ReadLine()
            If (strInput.Length > 1) Then
                strReturn = strInput
            End If
        Loop While (strReturn = String.Empty And Not bAllowEmpty)

        Return strReturn
    End Function

    Function ReadDevice(ByVal objGsm As Gsm)
        Dim strInput As String = String.Empty
        Dim strDevice As String = String.Empty
        Dim strPort As String = String.Empty
        Dim strMessage As String = String.Empty
        Dim strDevices As String = String.Empty
        Dim strSelectedDevice As String = String.Empty
        Dim i As Integer = 0
        Dim lsDeviceNames As List(Of String) = New List(Of String)

        strMessage = "Select a device: "

        strDevice = objGsm.FindFirstDevice()

        While (objGsm.LastError = 0)
            strDevices = strDevices & vbCrLf & "  " & i.ToString() & ": " + strDevice
            lsDeviceNames.Add(strDevice)
            strDevice = objGsm.FindNextDevice()
            i = i + 1
        End While

        strPort = objGsm.FindFirstPort()
        While (objGsm.LastError = 0)
            strDevices = strDevices & vbCrLf & "  " & i.ToString() & ": " + strPort
            lsDeviceNames.Add(strPort)
            strPort = objGsm.FindNextPort()
            i = i + 1
        End While

        If (strDevices = String.Empty) Then
            Console.WriteLine("No devices or COM ports where found.")
        Else
            Console.WriteLine(strMessage & strDevices)
            While (strSelectedDevice = String.Empty)

                Console.Write("  > ")
                strInput = Console.ReadLine()

                Dim iDevice As Integer

                If (Not Integer.TryParse(strInput, iDevice)) Then
                    iDevice = 0
                End If

                If (iDevice > lsDeviceNames.Count - 1) Then
                    iDevice = 0
                End If
                strSelectedDevice = lsDeviceNames(iDevice)
            End While
            Console.WriteLine("  Selected device: " & strSelectedDevice & vbCrLf)
        End If
        Return strSelectedDevice
    End Function
End Module

File: Visual CSharp.NET/Console Applications/Sms/Gsm/GsmProgram.cs
//-----------------------------------------------------------------------
// <copyright file="GsmProgram.cs" company="Auron">
//     Copyright (c) Auron Software - www.auronsoftware.com
// </copyright>
//-----------------------------------------------------------------------

namespace GsmProgram
{
    using System;
    using System.Collections.Generic;
    using System.IO;
    using System.Threading;
    using AxSms;

    class GsmProgram
    {       
        static void Main(string[] args)
        {
            AxSms.Gsm objGsm = new AxSms.Gsm();
            AxSms.Message objSmsMessage = new AxSms.Message();
            AxSms.Constants objSmsConstants = new AxSms.Constants();            

            // A license key is required to unlock this component after the trial period has expired.
            // Assign the LicenseKey property every time a new instance of this component is created (see code below). 
            // Alternatively, the LicenseKey property can be set automatically. This requires the license key to be stored in the registry.
            // For details, see manual, chapter "Product Activation".
            /*
            objGsm.LicenseKey = "XXXXX-XXXXX-XXXXX";
            */

            Console.WriteLine("Auron SMS Component {0}\nBuild: {1}\nModule: {2}\nLicense Status: {3}\nLicense Key: {4}\n", 
                objGsm.Version, objGsm.Build, objGsm.Module, objGsm.LicenseStatus, objGsm.LicenseKey);

            // Set Logfile (optional, for debugging purposes)
            objGsm.LogFile = Path.GetTempPath() + "Gsm.Log";
            Console.WriteLine("Log file used: {0}\n", objGsm.LogFile);

            string strDeviceName;
            if ((strDeviceName = InputDeviceName(objGsm)) == string.Empty)
            {
                Console.WriteLine("Press <ENTER> to continue.");
                Console.ReadLine();
                return;
            }

            objGsm.Open(strDeviceName, string.Empty, 0);
            Console.WriteLine("Open, result: " + objGsm.LastError + " (" + objGsm.GetErrorDescription(objGsm.LastError) + ")");
            if (objGsm.LastError == 36101) // 36101 means: modem requires pin code, see also: www.auronsoftware.com/kb/api-error-codes/
            {
                objGsm.Open(strDeviceName, ReadInput("Enter PIN code", false), 0);
                Console.WriteLine("Open, result: " + objGsm.LastError + " (" + objGsm.GetErrorDescription(objGsm.LastError) + ")");       
            }

            if (objGsm.LastError != 0)
            {
                Console.WriteLine("Press <ENTER> to continue.");
                Console.ReadLine();
                return;
            }

            // Message settings
            objSmsMessage.ToAddress = ReadInput("Enter recipient: ", false);
            objSmsMessage.Body = ReadInput("Enter message body: ", false);
            
            // Send the message
            objGsm.SendSms(objSmsMessage, objSmsConstants.MULTIPART_ACCEPT, 0);
            Console.WriteLine("SendSms, result: " + objGsm.LastError + " (" + objGsm.GetErrorDescription(objGsm.LastError) + ")");

            // Make sure the port gets closed
            objGsm.Close(); 
            Console.WriteLine("Ready.");
            Console.WriteLine("Press <ENTER> to continue.");
            Console.ReadLine();
        }

        static private string ReadInput(string strTitle, bool bAllowEmpty)
        {
            string strInput, strReturn = string.Empty;

            Console.WriteLine(strTitle);
            do
            {
                Console.Write("  > ");
                strInput = Console.ReadLine();
                if (strInput.Length > 1) strReturn = strInput;
            } while (strReturn == string.Empty && !bAllowEmpty);

            return strReturn;
        }

        static private string InputDeviceName(AxSms.Gsm objGsm)
        {
            string strInput, strDevice, strPort = string.Empty;
            string strMessage = string.Empty;
            int i = 0;
            var lsDeviceNames = new List<string>();

            strMessage += "Select a device: ";

            // Build a list of TAPI devices
            strDevice = objGsm.FindFirstDevice();
            while (objGsm.LastError == 0)
            {
                strMessage += "\r\n  " + (i).ToString() + ": " + strDevice;
                lsDeviceNames.Add(strDevice);
                strDevice = objGsm.FindNextDevice();
                i++;
            }

            // Build a list of COM ports
            strPort = objGsm.FindFirstPort();
            while (objGsm.LastError == 0)
            {
                strMessage += "\r\n  " + (i).ToString() + ": " + strPort;
                lsDeviceNames.Add(strPort);
                strPort = objGsm.FindNextPort();
                i++;
            }

            strDevice = string.Empty;

            if (lsDeviceNames.Count == 0)
                Console.WriteLine("No devices or COM ports where found.");             
            else
            {
                Console.WriteLine(strMessage);

                while (strDevice == string.Empty)
                {
                    Console.Write("  > ");
                    strInput = Console.ReadLine();

                    int iDevice = int.TryParse(strInput, out iDevice) ? iDevice : 0;
                    if (iDevice > lsDeviceNames.Count - 1) iDevice = 0; 
                    strDevice = lsDeviceNames[iDevice];
                }
                Console.WriteLine("  Selected device: " + strDevice + "\r\n");
            }
            return strDevice;
        }
    }
}
